source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+uses a ``Consumer`` functional interface to print the results below:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start findAllFiles
+   :end-before: end findAllFiles
+
+The next code example shows you how to retrieve and print the file names
+for all files that match the metadata fields specified in the query
+filter. The example also calls ``sort()`` and ``limit()`` on the returned
+``GridFSFindIterable`` to specify the order and maximum number of results:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start findMatchingFiles
+   :end-before: end findMatchingFiles
+
+Since ``metadata`` is an embedded document, the query filter specifies the
+``type`` field within the document using dot notation. See the server manual
+guide on how to :manual:`Query on Embedded/Nested Documents </tutorial/query-embedded-documents/>`
+for more information.
+
+For more information on the classes and methods mentioned in this section,
+see the following links:
+
+- :java-docs:`GridFSFindIterable <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSFindIterable.html>` API documentation
+- :java-docs:`GridFSBucket.find() <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBucket.html#find(org.bson.conversions.Bson)>` API documentation
+- :doc:`Sort Results </fundamentals/crud/read-operations/sort>`
+- :doc:`Limit the Number of Returned Results </fundamentals/crud/read-operations/limit>`
+
+.. _gridfs-download-files:
+
+Download Files
+--------------
+
+You can download a file from GridFS directly to a stream or you can save it
+to memory from a stream. You can specify the file to retrieve using either
+the file id or file name.
+
+File Revisions
+~~~~~~~~~~~~~~
+
+When your bucket contains multiple files that share the same file name,
+GridFS chooses the latest uploaded version of the file by default. To
+differentiate between each file that shares the same name, GridFS assigns
+files that share the same filename a revision number, ordered by upload time.
+
+The original file revision number is "0" and the next most recent file
+revision number is "1". You can also specify negative values which
+correspond to the recency of the revision. The revision value "-1"
+references the most recent revision and "-2" references the next most
+recent revision.
+
+The following code snippet shows how you can specify the second revision
+of a file in an instance of ``GridFSDownloadOptions``:
+
+.. code-block:: java
+
+   GridFSDownloadOptions downloadOptions = new GridFSDownloadOptions().revision(1);
+
+For more information on the enumeration of revisions, see the API
+documentation for :java-docs:`GridFSDownloadOptions <apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSDownloadOptions.html>`.
+
+Download a File to an Output Stream
```

Comment:
This might get a little too wordy, but I think it would be easier to understand the difference between this section and the next section if the titles referred to the use case for input/output streams instead of just the type of stream. For instance:

"Write GridFS Data to a File Using an Output Stream"
and
"Write GridFS Data to Application Memory Using an Input Stream"

Reply:
I thought about this comment for a while. I wonder if my assumption that working with input and output streams are understood when working with reading/writing files might not be suitable for new developers. I'm going to pass on this for now and try to think of a way to introduce streams more clearly.